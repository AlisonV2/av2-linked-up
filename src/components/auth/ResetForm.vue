<template>
  <form class="auth-form" @submit.prevent="resetPassword">
    <div class="row">
      <div class="col-12 text-center mb-4">
        <h2 class="title">Reset Password</h2>
      </div>
    </div>
    <div class="form-floating mb-3">
      <input
        type="email"
        class="form-control"
        placeholder="Please enter your email"
        v-model="email"
        required
      />
      <label>Email</label>
    </div>
    <div class="error mb-3 mt-3">{{ error }}</div>
    <div class="btn-center">
      <app-button>Reset password</app-button>
    </div>
    <slot />
  </form>
</template>

<script>
import * as Sentry from '@sentry/vue';

/**
 * @exports ParticipationForm
 * @type {Component}
 * @requires Sentry
 * @vue-data {string} email - v-model
 * @vue-event resetPassword
 */
export default {
  data() {
    return {
      email: '',
      error: null,
    };
  },
  methods: {
    resetPassword() {
      this.$store
        .dispatch('resetPassword', this.email)
        .then(() => {
          this.$router.push({ name: 'Login' });
        })
        .catch((err) => {
          this.error = err.message;
          Sentry.captureException(err);
          return;
        });
    },
  },
};
</script>

<style lang="scss" scoped>
.title {
  color: $dark;
  font-size: 1.8rem;
  margin-bottom: 1.5rem;
}
</style>
