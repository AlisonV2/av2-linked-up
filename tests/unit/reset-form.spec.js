import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';
import ResetForm from '@/components/auth/ResetForm.vue';
import AppTitle from '@/components/app/Title.vue';
import AppButton from '@/components/app/Button.vue';

let actions = {
  resetPassword: jest.fn(),
};

const store = createStore({
  actions,
});

const $router = {
  push: jest.fn(),
};

const config = {
  data() {
    return {
      email: 'test@test.com',
      error: null,
    };
  },
  global: {
    plugins: [store],
    mocks: {
      $router,
    },
    components: {
      'app-title': AppTitle,
      'app-button': AppButton,
    },
  },
};

/**
 * @module ResetFormTest
 */
describe('ResetForm.vue', () => {
  it('ResetForm Snapshot', () => {
    const component = shallowMount(ResetForm, config);
    expect(component.element).toMatchSnapshot();
  });

  it('Dispatches resetPassword action on submit', async () => {
    const component = shallowMount(ResetForm, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.resetPassword).toHaveBeenCalled();
  });

  it('Expects router push to have been called with Login params', async () => {
    const component = shallowMount(ResetForm, config);
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect($router.push).toHaveBeenCalledWith({ name: 'Login' });
  });

  it('Expects error message to be empty', () => {
    const component = shallowMount(ResetForm, config);
    const error = component.find('.error');
    expect(error.text()).toBeFalsy();
  });

  it('Checks if error message is correctly rendered', () => {
    let data = {
      email: '',
      error: 'Email is required',
    };
    const component = shallowMount(ResetForm, {
      ...config,
      data() {
        return data;
      },
    });
    const error = component.find('.error');
    expect(error.text()).toBe(data.error);
  });

  it('Expects an error to be thrown', async () => {
    actions = {
      resetPassword: () => {
        throw new Error('Error');
      },
    };
    const component = shallowMount(ResetForm, {
      data() {
        return {
          email: '',
          error: null,
        };
      },
      global: {
        plugins: [store],
        mocks: {
          $router,
        },
        components: {
          'app-title': AppTitle,
          'app-button': AppButton,
        },
      },
    });
    component.find('form').trigger('submit.prevent');
    await component.vm.$nextTick();
    expect(actions.resetPassword).toThrow('Error');
  });
});
