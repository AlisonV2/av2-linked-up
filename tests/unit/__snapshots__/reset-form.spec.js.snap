// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetForm.vue ResetForm Snapshot 1`] = `
<form
  class="auth-form"
>
  <div
    class="row"
  >
    <div
      class="col-12 text-center mb-4"
    >
      <h2
        class="title"
      >
        Reset Password
      </h2>
    </div>
  </div>
  <div
    class="form-floating mb-3"
  >
    <input
      class="form-control"
      placeholder="Please enter your email"
      required=""
      type="email"
    />
    <label>
      Email
    </label>
  </div>
  <div
    class="error mb-3 mt-3"
  />
  <div
    class="btn-center"
  >
    <app-button-stub />
  </div>
  
  
</form>
`;
